i6_models/primitives/specaugment.py

```diff
@@ -0,0 +1,118 @@
+
+    z = -torch.log(-torch.log(torch.rand((batch_dim, tensor.shape[axis])).to(device=tensor.device)))  # [B,dim]
+    _, indices = torch.topk(z, num_masks.max().item(), dim=1)
+
+    for i in range(num_masks.max().item()):
+        tensor = _mask(tensor, batch_axis, axis, indices[:, i], max_len)
+    return tensor
+
+
+def zero_specaugment(
+    audio_features: torch.Tensor,
+    time_min_num_masks: int,
+    time_max_num_masks: int,
+    time_mask_max_size: int,
+    freq_min_num_masks: int,
+    freq_max_num_masks: int,
+    freq_mask_max_size: int,
+):
+    """
+    Specaugment from legacy rossenbach/zeineldeen/zeyer attention setups (usually called specaugment_v2.py or so),
+    but without any step-based scheduling. Fills masks with zeros.
+
+    Basically just a convenience wrapper around _random_mask.
+
+    :param audio_features: e.g. log-mel features as [B, T, F]
+    :param time_min_num_masks: minimum number of masks along T
+    :param time_max_num_masks: maximum number of masks along T
+    :param time_mask_max_size: maximum size of masks along T
+    :param freq_min_num_masks: minimum number of masks along F
+    :param freq_max_num_masks: maximum number of masks along F
+    :param freq_mask_max_size: maximum size of masks along F
+    :return: masked audio features
+    """
+    assert len(tensor.shape) == 3
+    tensor = _random_mask(
+        audio_features, 0, 1, time_min_num_masks, time_max_num_masks, time_mask_max_size
+    )  # time masking
+    tensor = _random_mask(
+        audio_features, 0, 2, freq_min_num_masks, freq_max_num_masks, freq_mask_max_size
+    )  # freq masking
+    return tensor
+
+
+def zero_specaugment_by_length(
+    audio_features: torch.Tensor,
+    time_mask_per_n_frames: int,
+    time_min_num_masks: int,
+    time_mask_max_size: int,
+    freq_min_num_masks: int,
+    freq_max_num_masks: int,
+    freq_mask_max_size: int,
+):
+    """
+    Convenience wrapper around zero_specaugment with time-length adaptive number of masks
+
+    :param audio_features: e.g. log-mel features as [B, T, F]
+    :param time_mask_per_n_frames: maximum number of masks depending on length T.
+        They are still drawn depending on the full batch length, so shorter sequences
+        might get more masks than that by chance, or none at all when all masks
+        fall into the padding space.
+    :param time_min_num_masks: minimum number of masks along T
+    :param time_mask_max_size: maximum size of masks along T
+    :param freq_min_num_masks: minimum number of masks along F
+    :param freq_max_num_masks: maximum number of masks along F
+    :param freq_mask_max_size: maximum size of masks along F
+    :return: masked audio features
+    """
+    return zero_specaugment(
+        audio_features,
+        time_min_num_masks=time_min_num_masks,
+        time_max_num_masks=audio_features.size(1) // time_mask_per_n_frames,
```

Comment:
Is it somewhere guaranteed that `min_num_masks is <= max_num_masks` ?
What happens if it is not? (e.g. `torch.randint` fails?)

Reply:
I will add an assert!